File: src/983. Minimum Cost For Tickets/main.py
from typing import List
from array import array

class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        dp = array('H', [0] * 366)

        days = set(days)

        for day in range(1, 366):
            if day in days:
                first_ticket = max(dp[day-1] + costs[0], 0)
                second_ticket = max(dp[day-7] + costs[1], 0)
                third_ticket = max(dp[day-30] + costs[2], 0)

                dp[day] = min(first_ticket, second_ticket, third_ticket)
            else:
                dp[day] = dp[day-1]
        
        return dp[365]

File: src/983. Minimum Cost For Tickets/main.rs
use std::collections::HashSet;

impl Solution {
    pub fn mincost_tickets(days: Vec<i32>, costs: Vec<i32>) -> i32 {
        let days: HashSet<i32> = days.iter().cloned().collect();
        let mut dp = vec![0; 366];

        for day in 1..366 {
            let i = day as usize;
            if !days.contains(&day) {
                dp[i] = dp[i-1];
            } else {
                let c1 = (day - 1) as usize;
                let c2 = (day - 7).max(0) as usize;
                let c3 = (day - 30).max(0) as usize;

                dp[i] = (dp[c1] + costs[0]).min(dp[c2] + costs[1]).min(dp[c3] + costs[2]);
            }
        }

        dp[365]
    }
}